```{r}
library(flexdashboard)
```
# About 
## Content {data-width=50}
#### {.opening-page-text data-height=250}
<p style="font-family: 'Righteous', cursive; font-size: 80px; text-align: center;">KinAce</p>
Welcome to the KinAce web app for exploring the network of <b>kinase-substrate interactions</b> in the human genome. 

KinAce is jointly maintained by [Pandey](https://gpandeylab.org/) and [Schlessinger](http://www.schlessingerlab.org/) labs at [Icahn School of Medicine at Mt. Sinai, New York](https://icahn.mssm.edu/). 

For more information, contact [John Sekar](mailto:john.sekar@mssm.edu).

<h4 style="text-align: center;">How do I use the site?</h4>
KinAce provides different ways to examine the <i>functional</i> context of kinase-substrate interactions. To examine interactions

* of individual proteins, use the <b>Proteins</b> tab,
* between proteins in a curated KEGG pathway, use the <b>Pathways</b> tab,
* of proteins enriched in specific InterPro domains, use the <b>Domains</b> tab,
* in a custom gene set, use the <b>Input</b> tab.

Each tab displays a network of proteins, both kinases (colored by kinase group) and non-kinases (colored gray). Each interaction is indicated by a directed edge from the respective kinase to the substrate. Try...

* interacting with the network by moving nodes around, zooming and panning,
* selecting an edge and following the displayed links to the original sources,
* toggling the <u>Include adjacent proteins</u> and <u>Exclude disconnected proteins</u> options.

<h4 style="text-align: center;">Where is the data from?</h4>
KinAce aggregates interactions from [PhosphoSitePlus](https://www.phosphosite.org/), [iPTMnet](https://research.bioinformatics.udel.edu/iptmnet/) and [EPSD](https://epsd.biocuckoo.cn/). We also use kinase information from [KinHub](http://kinhub.org/), [Coral](http://phanstiel-lab.med.unc.edu/CORAL/) and the [Dark Kinase Knowledgebase](https://darkkinome.org/). Protein names and gene symbols are from [UniProt](https://www.uniprot.org/uniprotkb?query=human&facets=reviewed%3Atrue%2Cmodel_organism%3A9606) and [HGNC](https://www.genenames.org/) respectively. Gene set enrichment analysis for [KEGG](https://www.genome.jp/kegg/pathway.html) and [InterPro](https://www.ebi.ac.uk/interpro/) terms was performed using [Enrichr](https://maayanlab.cloud/Enrichr/).
<br>

<h4 style="text-align: center;">How do I download the data?</h4>
On each tab, use the <u>Export</u> options to download the visualized network as a table (CSV) or in selected graph visualization formats (DOT, GML, GraphML). To download the entire dataset, click <a href="./data/2023-10-30-kinace-dataset.zip" download>here</a>.


## Image {data-width=20}
### {}

```{r}
valueBoxOutput('vbox_kinases')
output$vbox_kinases <- renderValueBox({
  valueBox(value=n_kinases,caption=tags$p("Kinases",class="opening-page-text"),icon="fa-solid fa-k")
  })
```
### {}
```{r}
valueBoxOutput('vbox_nonkinases')
output$vbox_nonkinases <- renderValueBox({
  valueBox(value=n_nonkinases,caption=tags$p("Non-kinases",class="opening-page-text"),icon="fa-solid fa-s")
  })
```
### {}
```{r}
valueBoxOutput('vbox_interactions')
output$vbox_interactions <- renderValueBox({
  valueBox(value=n_interactions,caption=tags$p("Interactions",class="opening-page-text"),icon="fa-solid fa-i")
  })
```
### {}
```{r}
valueBoxOutput('vbox_kki')
output$vbox_kki <- renderValueBox({
  valueBox(value=n_kki,caption=tags$p("Kinase-kinase interactions",class="opening-page-text"),icon="fa-solid fa-kip-sign")
  })
```
### {}
```{r}
valueBoxOutput('vbox_auto')
output$vbox_auto <- renderValueBox({
  valueBox(value=n_auto,caption=tags$p("Autophosphorylations",class="opening-page-text"),icon="fa-solid fa-a")
  })
```

##### Word cloud of pathway names enriched in kinase substrates
![](img/wordcloud.png){width="400px"}

##### Flow diagram of kinase-kinase interaction data, organized by kinase-group
![](img/sankey.png){width="400px"}



