# About 

## Content {data-width=50}
### {.opening-page-text}
<p style="font-family: 'Righteous', cursive; font-size: 80px;">KinAce</p>
Welcome to the KinAce visualization portal! Our goal is to visualize the network of <b>kinase-substrate interactions</b> in the human genome. Click on the tabs above to:

* examine the interactions of <b>individual proteins</b> to other kinases and substrates,
* visualize kinase-substrate interactions in <b>curated pathways</b>,
* import a <b>custom gene set</b> and visualize its kinase-substrate interactions.

In each of these visualizations, proteins are represented as nodes and interactions are represented as directed edges from the kinase to the substrate. Use the mouse to select nodes and edges, move nodes, pan and zoom into the network. There are additional controls in the sidebar to highlight kinase groups and customize the selection of proteins on display. When you click on an edge, information about that interaction will be displayed on the right. 

<h3>FAQ</h3>
<h4>Where is the data from?</h4>
We aggregate interactions from [PhosphoSitePlus](https://www.phosphosite.org/), [iPTMnet](https://research.bioinformatics.udel.edu/iptmnet/) and [EPSD](https://epsd.biocuckoo.cn/). 
<h4>Who maintains the site?</h4>
The portal is jointly maintained by [Pandey](https://gpandeylab.org/) and [Schlessinger](http://www.schlessingerlab.org/) labs at [Icahn School of Medicine at Mt. Sinai, New York](https://icahn.mssm.edu/). For more information, contact [John Sekar](mailto:john.sekar@mssm.edu).

## Image {data-width=30}
##### Word cloud of names of pathways enriched in kinases and their substrates
![](img/wordcloud.png){width="400px"}

##### Flow diagram of kinase-kinase interactions (left-kinase, right-targeted kinase), organized by kinase-group
![](img/sankey.png){width="400px"}



