
# About 

## Image {data-width="300px"}
### {}


## Content
##### Introduction
The KSI Visualization Portal enables graphical visualization of publicly-available data on interactions between protein kinases and their substrates in the human genome. Currently, our sources include [PhosphoSitePlus](https://www.phosphosite.org), [iPTMNet](https://research.bioinformatics.udel.edu/iptmnet/) and [EPSD](http://epsd.biocuckoo.cn/), cross-referenced with reviewed entries from [UniProt](https://www.uniprot.org/).

##### Examine a protein's kinase-substrate interactions
The "Examine a protein" tab lets you select a single protein and visualize its interactions, whether as the kinase or as the substrate. 

##### Visualize cellular pathways
The "Visualize pathways" tab lets you select a cellular pathway (from [KEGG](https://www.genome.jp/kegg/pathway.html) database) and visualize the kinase-substrate interactions that are available about proteins in that pathway.

##### Visualize custom gene sets
The "Visualize custom gene sets" tab lets you examine the kinase-substrate interactions on any input gene-set.

##### Information about sources
On each of our visualizations, clicking on a particular interaction displays more information about it: which two proteins are involved, what sites are relevant and which sources document them.

##### Navigating the interaction network
On the "Examine a protein" tab, you can walk through the interaction network by double-clicking on a different protein, which re-centers the visualization. You can also double-click on a protein on any other tab and click on the "Examine a protein" tab to examine its interactions.

##### Contact
The KSI Visualization portal is maintained by [Pandey lab](https://gpandeylab.org/) at Mt. Sinai School of Medicine, Dept. Genetics and Genomic Sciences.